Enhance "mail" module

Undefined-Variables opened the pull request:
Done correctly.

Undefined-Variables commented:
@pH-7 I'm not expecting you to merge this as it serves as more of an example of the idea. Maybe you can have a look and come up with something similar. Please test yourself if you do merge as this was put together with showing my thoughts on the messaging system.

Do you see how I include search and forms all on a single page? Think this really does a lot for the navigational experience for users.

pH-7 commented:
>  showing my thoughts on the messaging system in mind.

Brilliant @Undefined-Variables!
I will checkout out your branch and review it as well. 
There are several changes in there! Not only the weird messages fixed. 
I know that the "mail" module is quite outdated. I really appreciate you updated some styles and the menu in there. 

I will be back here very soon once I reviewed everything.

pH-7 commented on _protected/app/system/modules/mail/controllers/MainController.php:
```diff
@@ -37,6 +37,9 @@ class MainController extends Controller
     private $iProfileId;
+    
+    /** @var int */
+    private $iSender;
```
FYI, this attribute can be removed, since it isn't used anywhere in the class.

Undefined-Variables replied:
Yes I forgot to remove it, I started off with an idea of how I was going to get the senders info but changed my mind and never removed it. 

I think removing the trash all together is a good idea. Simply mark as deleted until both users have deleted at which point it can be dropped form the database. 

Have you viewed it via a browser? You will understand why I removed the echo where, by and sort if you do. I was unable to pass the sender and subject to the mailform even though it is displayed on the same page and couldn't pass values in the use of the mailform() function either. I need to think on this a little more.

pH-7 commented on _protected/app/system/modules/mail/forms/SearchMailForm.php:
```diff
@@ -19,13 +19,35 @@ public static function display()
+        $oForm->addElement(new \PFBC\Element\HTMLExternal('<div class="msg-search-input">'));
         if (!self::isAdminLoggedAndNotUser()) {
-            $oForm->addElement(new Select(t('Where:'), 'where', [MailModel::INBOX => t('Inbox'), MailModel::OUTBOX => t('Outbox'), MailModel::TRASH => t('Trash')]));
+            $oForm->addElement(new Select((''), 'where', [MailModel::INBOX => t('Inbox'), MailModel::OUTBOX => t('Outbox'), MailModel::TRASH => t('Trash')]));
```
Was it on purpose that you omitted the `t('Where:')` string?

Undefined-Variables replied:
This where: is being echo'd in quotes because its being used for the label of the form element. The 'where' which is the second parameter is the MySQL clause. Have a look in a browser you will understand what I mean.

pH-7 commented on _protected/app/system/modules/mail/controllers/MainController.php:
```diff
@@ -251,10 +260,10 @@ public function trash()
 
             if (empty($oMail)) {
-                $this->sTitle = t('Sorry!');
+                $this->sTitle = t('Not Found!');
```
Much better message! Thanks!

Undefined-Variables replied:
I just changed a few of the messages like this as we are saying sorry for no reason :) 

Also named the page titles as messages rather than mail as mail is what a postman brings haha. and email is in no way related to pm's :)

Undefined-Variables commented:
There is an example you can see here

https://undefined-variables.com/ph7cms/

pH-7 commented:
> There is an example you can see here
> 
> https://undefined-variables.com/ph7cms/
> 
> [user1@user1.com](mailto:user1@user1.com)
> [user2@user2.com](mailto:user2@user2.com)
> 
> passwd is user123 for both accounts and messages between them so you can see it in action.

<img width="703" alt="Screenshot 2020-04-05 at 1 20 00 pm" src="https://user-images.githubusercontent.com/1325411/78466152-7d50de80-76ed-11ea-9fa7-093befe478db.png">

Wow, I wasn't expecting this. That's so nice of you. It the job much easier :)

pH-7 commented on _protected/app/system/modules/mail/views/base/tpl/main/msg.inc.tpl:
```diff
@@ -14,19 +23,56 @@
-    <div><a href="{{ $design->url('mail','main','compose',"$usernameSender,$subject") }}">{lang 'Reply'}</a> | {{ LinkCoreForm::display($label_txt, 'mail', 'main', $set_to, array('id'=>$msg->messageId)) }}
-    {if $is_trash} | {{ LinkCoreForm::display(t('Move to Inbox'), 'mail', 'main', 'setrestore', array('id'=>$msg->messageId)) }}{/if}</div>
-  </div>
+        {{ MailForm::display() }}
```
We will probably have to change that form. Otherwise, we get the recipient and subject fields as well when we reply.

Undefined-Variables replied:
Yeah I couldn't pass the username of the sender or the subject to the form. I think its because its being called before the functions in the controller then the model have had enough time to return the values. The form is there instantly the rest of the process is still taking place.

I just noticed if you are the sender you can delete messages from the list but if you are recipient it shows move to inbox. I will have a look when I'm able to.

Did you like the idea of the search being there on the page? I have done this with photo's and Video's too on my live site.

Undefined-Variables replied:
How would you change the form? Remove the recipient and subject fields as a reply should hold the subject with one Re: preceding it but I don't see how the info can be added before the form is called? Maybe in the model is the ideal place?

pH-7 commented on templates/themes/base/tpl/top_menu.inc.tpl:
```diff
@@ -167,20 +167,11 @@
-                <li><a href="{{ $design->url('mail','main','search') }}" title="{lang 'Find Messages'}"><i class="fa fa-search"></i> {lang 'Search'}</a></li>
-              </ul>
+            <li>
+               <a href="{{ $design->url('mail','main','inbox') }}" title="{lang 'Inbox'}"><i class="fa fa-inbox"></i> {lang 'Messages'} {if $count_unread_mail}<span class="badge">{count_unread_mail}</span>{/if} <span class="caret"></span></a>
```
BTW, if you decide to remove the drop-down sub menus, it's a good idea to get rid of `<span class="caret"></span>`.  The 'caret' usually indicates that the link is a drop-down one, but it's not the case anymore.

Undefined-Variables replied:
Oh yeah. Thats a silly thing to miss :(

pH-7 commented on _protected/app/system/modules/mail/views/base/tpl/main/msg.inc.tpl:
```diff
@@ -14,19 +23,56 @@
+    {/each}
+    {/if}
+        <hr>
+       
```
Since we display up to the 25 messages of the conversation before the 'reply' form, it might be a good idea to directly scroll down the page to the bottom (to the reply form). 
It's probably more "user friendly".

We can do it easily with jQuery like,
```js
    var sHash = location.hash.substr(1);
    var $target = $('html, body');
    $target.animate({scrollTop: $target.height()}, 1000);
```

Undefined-Variables replied:
Or you could put the reply at the top and and sort ascending instead. I was thinking of using offset and pages of history but just seems unnecessary.

Undefined-Variables replied:
That jQuery animate is a nice touch. 

And I have favicon, menu and inbox(2) all telling me how many unread messages I have.

pH-7 replied:
> Or you could put the reply at the top and and order by ASC instead. I was thinking of using offset and pages of history but just seems unnecessary.

If you move the reply form to the top, it might be harder for users to read the previous messages before replying. I think having the reply form at the bottom is still a good thing

Undefined-Variables replied:
Yes I agree. I will find a way around the recipient and subject being added.

pH-7 commented on _protected/app/system/modules/mail/models/MailModel.php:
```diff
@@ -73,7 +73,29 @@ public function readSentMsg($iSender, $iMessageId)
+            WHERE msg.recipient = :recipient
+            AND msg.sender = :sender OR msg.sender = :recipient
+            AND NOT FIND_IN_SET(\'recipient\', msg.toDelete)
+            AND NOT FIND_IN_SET(\'sender\', msg.toDelete) LIMIT 25'
```
Probably need to add `ORDER BY sendDate DESC` here, before the LIMIT clause.

Undefined-Variables replied:
Yes you are right, good idea.

pH-7 commented on _protected/app/system/modules/mail/views/base/tpl/list.inc.tpl:
```diff
@@ -22,16 +33,21 @@
                 {{ $move_to = ($is_delete) ? 'delete' : 'trash' }}
                 {{ $label_txt = ($is_delete) ? t('Delete') : t('Trash') }}
 
+                {{ $subject = escape(str_replace('re ', '', $msg->title), true) }}
```
For your information, here you just replace the "re" lower case. It's case-sentitive. You will have to use `str_ireplace()` for case-insensitive replace https://www.php.net/manual/en/function.str-ireplace.php

Undefined-Variables replied:
Oh okay wow I didn't know that. I thought str_replace was for complex situations where you may need to use regex and preg was for more simple use.  I like it when I learn something like this 👍

pH-7 replied:
👍 
No, actually, it's the other way around. `preg_replace` with regex would be for more complex scenarios. When it's possible, it's always better to use str_(i)replace since it's more readable (regex aren't easy to read) and it's faster too.

pH-7 commented:
<img width="887" alt="Screenshot 2020-04-05 at 2 04 19 pm" src="https://user-images.githubusercontent.com/1325411/78466660-3bc33200-76f3-11ea-92ea-3c358d0583d5.png">

We probably have a design issue here...

Undefined-Variables commented:
> <img alt="Screenshot 2020-04-05 at 2 04 19 pm" width="887" src="https://user-images.githubusercontent.com/1325411/78466660-3bc33200-76f3-11ea-92ea-3c358d0583d5.png">
> 
> We probably have a design issue here...

Oh must be screen size need to use a media breakpoint.

pH-7 commented:
Looks like it doesn't browser tab favicon's alert isn't working when sending new messages, and doesn't display the amount of unread messages when you send new ones.

Undefined-Variables commented:
> Looks like it doesn't browser tab favicon's alert isn't working when sending new messages, and doesn't display the amount of unread messages when you send new ones.

Really? What browser are you using? Works with Chrome! Your not a FireFox person are you? I have had a few odd things with FireFox, the browser doesn't like me! I will have a look in both now.

Undefined-Variables commented:
The favicon is working on my localhost so must have missed something when I uploaded this copy. 

Also the str_replace only removes the re on localhost and leaves the RE. Wonder why that works.

![Capture](https://user-images.githubusercontent.com/52648055/78466895-bcd3f680-76fe-11ea-9644-28bc887a364a.PNG)

Undefined-Variables commented:
> Looks like it doesn't browser tab favicon's alert isn't working when sending new messages, and doesn't display the number of unread messages when you send new ones. Probably an issue with `MailCoreModel::countUnreadMsg()` At least, this occurs on your demo website :)

Does it work in the message menu next to inbox(0) ??? Uses same function. think its settings in my php.ini on demo site stopping the favicon being altered

Undefined-Variables commented:
![Capture](https://user-images.githubusercontent.com/52648055/78466996-0244f380-7700-11ea-9c1a-a6fc1326c416.PNG)

Undefined-Variables commented:
Working my end.

pH-7 commented:
> Also the str_replace only removes the re on localhost and leaves the RE. Wonder why that works.

You will need to use `str_ireplace()`, then the `RE` upper case will be gone. It's because the other one is case-sensitive

pH-7 commented:
> > <img alt="Screenshot 2020-04-05 at 2 04 19 pm" width="887" src="https://user-images.githubusercontent.com/1325411/78466660-3bc33200-76f3-11ea-92ea-3c358d0583d5.png">
> > We probably have a design issue here...
> 
> Oh must be screen size need to use a media breakpoint.

Yeap, probably some CSS tweaks and media queries needed there

Undefined-Variables commented:
> > Also the str_replace only removes the re on localhost and leaves the RE. Wonder why that works.
> 
> You will need to use `str_ireplace()`, then the `RE` upper case will be gone. It's because the other one is case-sensitive

Yes but we want to have the RE: on a reply before it was giving RE: on the first reply then every reply after had RE: re re re re adding a re on each message so eventually the actual subject was lost in the view due to only allowing the first 20 characters as a result of the substr used to limit the length.

Undefined-Variables commented:
> > > <img alt="Screenshot 2020-04-05 at 2 04 19 pm" width="887" src="https://user-images.githubusercontent.com/1325411/78466660-3bc33200-76f3-11ea-92ea-3c358d0583d5.png">
> > > We probably have a design issue here...
> > 
> > 
> > Oh must be screen size need to use a media breakpoint.
> 
> Yeap, probably some CSS tweaks and media queries needed there

I will make all your suggested changes and push again tomorrow :)

Undefined-Variables commented:
> Looks like the browser tab favicon's alert isn't working when sending new messages, and doesn't display the number of unread messages when you send new ones. Probably an issue with `MailCoreModel::countUnreadMsg()` At least, this occurs on your demo website :)

Sorry just read this again, it should only tell you the unread count if you are the recipient as it lands in your inbox. You don't want to be told you have unread sent messages do you?

Undefined-Variables commented:
I have just updated test site and my repository. I have done everything but cant seem to get this MailForm to work in the way I need. I will keep thinking but feel free to look. I'm not requesting the pull until everything is done that we discussed.

Undefined-Variables commented:
@pH-7 quick question. If the is online feature stopped working on a live site where would you look first? Not worked for a while and this is a big part of making a site feel active. Looks dead with nobody ever online.

pH-7 commented:
> I have just updated test site and my repository. I have done everything but cant seem to get this MailForm to work in the way I need. I will keep thinking but feel free to look. I'm not requesting the pull until everything is done that we discussed.

Once everything is done on your side, I will merge it in the dev branch and make sure everything is working as expected. Just let me know. Thanks again @Undefined-Variables!!! I REALLY appreciate your help

pH-7 commented:
> @pH-7 quick question. If the is online feature stopped working on a live site where would you look first? Not worked for a while and this is a big part of making a site feel active. Looks dead with nobody ever online.

Could you be more specific? Do you speak about the "User inactivity timeout" setting in the admin setting form?

pH-7 commented:
> > > > <img alt="Screenshot 2020-04-05 at 2 04 19 pm" width="887" src="https://user-images.githubusercontent.com/1325411/78466660-3bc33200-76f3-11ea-92ea-3c358d0583d5.png">
> > > > We probably have a design issue here...
> > > 
> > > 
> > > Oh must be screen size need to use a media breakpoint.
> > 
> > 
> > Yeap, probably some CSS tweaks and media queries needed there
> 
> I will make all your suggested changes and push again tomorrow :)

Sounds great! 🤩

pH-7 commented:
> > Looks like the browser tab favicon's alert isn't working when sending new messages, and doesn't display the number of unread messages when you send new ones. Probably an issue with `MailCoreModel::countUnreadMsg()` At least, this occurs on your demo website :)
> 
> Sorry just read this again, it should only tell you the unread count if you are the recipient as it lands in your inbox. You don't want to be told you have unread sent messages do you?

Yes indeed, sorry if I wasn't clear. I was speaking about the **recipient user**. I couldn't see the number of unread messages as the recipient.

Undefined-Variables commented:
Yeah is always a good idea to allow time for testing, I will be running it a live site where it will see heavy us, it will be using the same code only I have styled it to fit the theme so I will soon know if something is not right. I have done similar in other modules and not had an issue yet so im optimistic.

I'm not really sure what has happened to the online status of users, when I am on a test install or development site I can see ajax activity in the console setting user activity every couple of seconds but on the live site this is not happening at all. Every user has a black/grey status indicator and I have to be honest I am great with html, css, php, mysql but when javascript is involved, I have not used it much and every script is minimized so hard to read and not sure where to start looking. So what im asking is where is user activity monitored? Login must trigger the process right?

Undefined-Variables commented:
As unlikely as it seems asking that question triggered a thought that made me realise where this is and has to be initiated by setting the users session as logging in!

pH-7 commented:
> Yeah is always a good idea to allow time for testing, I will be running it on a live site where it will see some heavy us, it will be using the same code only I have styled it to fit the theme so I will soon know if something is not right. I have done similar in other modules and not had an issue yet so im optimistic.
> 
> I'm not really sure what has happened to the online status of users, when I am on a test install or development site I can see ajax activity in the console setting user activity every couple of seconds but on the live site this is not happening at all. Every user has a black/grey status indicator and I have to be honest I am great with html, css, php, mysql but when javascript is involved, I have not used it much and every script is minimized so hard to read and not sure where to start looking. So what im asking is where is user activity monitored? Login must trigger the process right?

When a user is logged in, the JS file triggers this one https://github.com/pH7Software/pH7-Social-Dating-CMS/blob/master/_protected/app/system/modules/user/assets/ajax/setActivityAjax.php#L20 which updates lastActivity DB field.

Then, `UserCoreModel::isOnline()` https://github.com/pH7Software/pH7-Social-Dating-CMS/blob/22e978e207566a756b5fb5a519c5beff8c82cd3e/_protected/app/system/core/models/UserCoreModel.php#L478-L492 checks ` lastActivity` DB field according to the current server time.

If you loop into it, `UserCoreModel::isOnline()` method has a second argument, which takes the number of minutes when a member becomes inactive (offline). That argument is used by the "User inactivity timeout" field located in the admin dashboard, Setting -> General -> Automation.

Hope it helps!

pH-7 commented:
By the way, concerning this PR changes, let me know once you are done with it, and so, I will be able to merge it and review everything on my localhost through my PhpStorm.

At the moment, GitHub says that the repository that submitted this pull request has been deleted...

Undefined-Variables commented:
> > Yeah is always a good idea to allow time for testing, I will be running it on a live site where it will see some heavy us, it will be using the same code only I have styled it to fit the theme so I will soon know if something is not right. I have done similar in other modules and not had an issue yet so im optimistic.
> > I'm not really sure what has happened to the online status of users, when I am on a test install or development site I can see ajax activity in the console setting user activity every couple of seconds but on the live site this is not happening at all. Every user has a black/grey status indicator and I have to be honest I am great with html, css, php, mysql but when javascript is involved, I have not used it much and every script is minimized so hard to read and not sure where to start looking. So what im asking is where is user activity monitored? Login must trigger the process right?
> 
> When a user is logged in, the JS file triggers this one https://github.com/pH7Software/pH7-Social-Dating-CMS/blob/master/_protected/app/system/modules/user/assets/ajax/setActivityAjax.php#L20 which updates lastActivity DB field.
> 
> Then, `UserCoreModel::isOnline()`
> 
> https://github.com/pH7Software/pH7-Social-Dating-CMS/blob/22e978e207566a756b5fb5a519c5beff8c82cd3e/_protected/app/system/core/models/UserCoreModel.php#L478-L492
> 
> checks ` lastActivity` DB field according to the current server time.
> If you loop into it, `UserCoreModel::isOnline()` method has a second argument, which takes the number of minutes when a member becomes inactive (offline). That argument is used by the "User inactivity timeout" field located in the admin dashboard, Setting -> General -> Automation.
> 
> Hope it helps!

You are Awesome @pH-7 you have no idea how much that has helped. I had most of it right but couldn't see how the 1 second made any sense but now I understand perfectly. I really appreciate you taking the time to explain that to me :) 

I had to put things on hold yesterday and deal with a server issue but all sorted now and I will finish this off this evening. 

Thank you!

Undefined-Variables commented:
I cant seem to update a repository from Github Desktop so I just delete it and fork then it sync's as it should but if I make changes in a working example it tells me 18 files have been altered and asks if I would like to push to my repository. Then if I click yes it piles 8000 files and crashes my PC!

Undefined-Variables commented:
> > Yeah is always a good idea to allow time for testing, I will be running it on a live site where it will see some heavy us, it will be using the same code only I have styled it to fit the theme so I will soon know if something is not right. I have done similar in other modules and not had an issue yet so im optimistic.
> > I'm not really sure what has happened to the online status of users, when I am on a test install or development site I can see ajax activity in the console setting user activity every couple of seconds but on the live site this is not happening at all. Every user has a black/grey status indicator and I have to be honest I am great with html, css, php, mysql but when javascript is involved, I have not used it much and every script is minimized so hard to read and not sure where to start looking. So what im asking is where is user activity monitored? Login must trigger the process right?
> 
> When a user is logged in, the `setUserActivity.js` JS file triggers this one https://github.com/pH7Software/pH7-Social-Dating-CMS/blob/master/_protected/app/system/modules/user/assets/ajax/setActivityAjax.php#L20 which updates lastActivity DB field.
> 
> Then, `UserCoreModel::isOnline()` checks `lastActivity` DB field with the current **server time**, in order to know if the user is still logged in.
> 
> https://github.com/pH7Software/pH7-Social-Dating-CMS/blob/22e978e207566a756b5fb5a519c5beff8c82cd3e/_protected/app/system/core/models/UserCoreModel.php#L478-L492
> 
> If you loop into it, `UserCoreModel::isOnline()` method has a second argument, which takes the number of minutes when a member becomes inactive (offline). That argument is used by the "User inactivity timeout" field located in the admin dashboard, Setting -> General -> Automation.
> 
> Hope it helps!

I have checked and all this is correct and I see ajax setting the activity in dev tools > network but all users have a grey status indicator still so this makes me think its a css problem, something must be overriding the css. I will look in to it further later.

Undefined-Variables commented:
It was missing the color as when I wrote my theme most of the default css sheets were obsolete but the status was .green in color.css which was no longer needed.

Undefined-Variables commented:
![green](https://user-images.githubusercontent.com/52648055/78632097-06a31500-7896-11ea-9e98-0d011dc83417.PNG)

pH-7 commented:
> It was missing the color as when I wrote my theme most of the default css sheets were obsolete but the status was .green in color.css which was no longer needed.



> ![green](https://user-images.githubusercontent.com/52648055/78632097-06a31500-7896-11ea-9e98-0d011dc83417.PNG)

So happy for you! Sometimes you can spend days trying to fix a very simple issue... :-)

pH-7 commented:
> > > Yeah is always a good idea to allow time for testing, I will be running it on a live site where it will see some heavy us, it will be using the same code only I have styled it to fit the theme so I will soon know if something is not right. I have done similar in other modules and not had an issue yet so im optimistic.
> > > I'm not really sure what has happened to the online status of users, when I am on a test install or development site I can see ajax activity in the console setting user activity every couple of seconds but on the live site this is not happening at all. Every user has a black/grey status indicator and I have to be honest I am great with html, css, php, mysql but when javascript is involved, I have not used it much and every script is minimized so hard to read and not sure where to start looking. So what im asking is where is user activity monitored? Login must trigger the process right?
> > 
> > 
> > When a user is logged in, the JS file triggers this one https://github.com/pH7Software/pH7-Social-Dating-CMS/blob/master/_protected/app/system/modules/user/assets/ajax/setActivityAjax.php#L20 which updates lastActivity DB field.
> > Then, `UserCoreModel::isOnline()`
> > https://github.com/pH7Software/pH7-Social-Dating-CMS/blob/22e978e207566a756b5fb5a519c5beff8c82cd3e/_protected/app/system/core/models/UserCoreModel.php#L478-L492
> > 
> > checks ` lastActivity` DB field according to the current server time.
> > If you loop into it, `UserCoreModel::isOnline()` method has a second argument, which takes the number of minutes when a member becomes inactive (offline). That argument is used by the "User inactivity timeout" field located in the admin dashboard, Setting -> General -> Automation.
> > Hope it helps!
> 
> You are Awesome @pH-7 you have no idea how much that has helped. I had most of it right but couldn't see how the 1 second made any sense but now I understand perfectly. I really appreciate you taking the time to explain that to me :)
> 
> I had to put things on hold yesterday and deal with a server issue but all sorted now and I will finish this off this evening.
> 
> Thank you!

You're very welcome!! 😊

Undefined-Variables commented:
Yeah I have spent days looking at this thinking it should work but never even thought about the css. I had almost given up and was considering a rebuild. Thank you for pointing me in the right direction

Undefined-Variables commented:
I’m installing phpstorm. I have to give this a go, I’ve been using Komodo IDE for 10yrs now but after you said that I googled it and it sounds like a dream

polynamaude commented:
@Undefined-Variables You can also use Eclipse with the PHP Development Toolkit (PDT). It's free and does much more than Komodo. Because that last one ain't really and IDE but more of a colored sentence viewer.

Undefined-Variables commented:
@polynamaude Thank you!

pH-7 commented:
FYI @Undefined-Variables @polynamaude, if this can help... if you prove that you are fully committed to an open source project (like pH7CMS or any others), PhpStorm gives the license for free https://www.jetbrains.com/community/opensource/#support
It works and they are quite responsive.

In the meantime, it doesn't cost anything to use their 30 days trial (you can still use it after the trial but for just 30 or 60 minutes per time).

polynamaude commented:
@pH-7 I know this trick about PHPStorm but I don't like much their business model and it's mostly opposite of being OpenSource.

Years ago we had to choose a official IDE for the TikiWikiCMS. I had the chance to see the advantages of PHPStorm. We still chosen Eclipse because it's free without limitation of use.

polynamaude commented:
@pH-7 phpStorm act like Apple computer who'd give computer to school if they'd be allowed to make a presentation to parents.

Undefined-Variables commented:
@pH-7 Thanks for the info. @polynamaude I had a look at both of these editors and TBH I found both hard to navigate but I suppose if you are used to using a particular IDE its a learning curve no matter what you do. I wasted so much time yesterday trying to figure these out so for now I am sticking to what I know and when I have time I will look again. Komodo IDE works for me mainly because I know it inside out and understand it.

I read a review that said "If you are developing in PHP and don't use phpstorm you are working to hard" this is what made me want to try it but it slowed me down trying to learn how it works :/

Undefined-Variables commented:
Still working on this, just had a busy couple of days but I have not forgotten :)

Undefined-Variables commented:
@polynamaude Have a word with pool man! He needs a few hard poly words I think, he is doing me in!

Undefined-Variables commented:
@polynamaude he blocked me haha, really didn't like the truth.

polynamaude commented:
@polynamaude Hi !
You should try out Visual Studio Code. Has nothing to do with the standard visual studio we are used to. It's pretty light weight, based on the Atom IDE.
There's all sort of nice plugins for web development. You can have PHP,JS,CSS/HTML validation. It works really nice with **git** and manage automagicaly the tags you used (for example the TODO tag can be directly linked to issues on your **git**). Sometime it's worth it to take a day or two, even a week to learn new tools that will allow you to work better. For example the three IDE that we talked about (VSCode, PHPStorm, Eclipse) all manage tags, issues on your git, and so much more. I help you to do things better...

I saw many very good programmer getting fired not because they didn't know their subject but because they had problem fitting with others. Pretty often people learn computing "on the fly" without learning the base of project management. But it goes further...

I'll give you the example of **Design Pattern**. There's one called MVC and it's somewhat used in this CMS. There's also one called **Decorator** and **Singleton**. That last one is used pretty much in all CMS and it's used for our main object in this one. By following the rules of separation you guarantee that your modifications will continue to work when the code is upgraded. Because in a team, people work on different part and it's important to respect this.

What's the link with project management ? Well those IDE also helps you do better documentation because it's the most important part of API (or source code in general). People like to see result and don't like to plan stuff. But we can save lot of time by doing so. Save us from doing modification later on because we forgot to separate some part of our code. The main objective of all those separation is a two-time gain. First it allow parts of code to work with each other with defined limits. This also allow upgrade to go on without breaking. (forward compatibility). Also it allow easier modification when needed.

For example all the work done on the back-end can work with any template the user could design as long he (or her) follow the rules set forth in our template (for example what variables are available). This show where code shall go.

A week used to learn stuff that can save you months of time is always a good deal ;-)

PM

Undefined-Variables commented:
Not sure what talking to yourself is all about @polynamaude but anyway... I have tried  Visual Studio Code and yes I have to agree its quite nice to use, light weight and has tons of add-on's and fully supports multiple languages but I couldn't find decent remote ftp add-on for it, the best available is ftp-remote and hasn't got any nice way of displaying your remote servers file system like a tree for example. Every time you need a file you have to F1 and navigate directories one at a time. It makes things long winded especially when Komodo and even Atom have nice trees to navigate. 

I tried Atom for about 8Mths then it just stopped opening and I reinstalled and got advice on their forum but nothing worked, so went back to Komodo IDE and realized just how bad Atom is compared to Komodo IDE, I returned to the forum and asked for advice and what I would make it better and my post was hidden so they didn't like constructive criticism. Great software only becomes great by people saying "hey this part is not perfect" and it gets improved, they didn't want to engage in this talk!

I fully understand how MVC works, I first came across it in a project called "Bitwasp" it was written by @afk11 and was very nicely written, clean and a great example of this style of writing code. I then played with CodeIgnighter  for a few months and have since used it to code two sites for friends so fully appreciate the logic of it but what I don't get is if the better IDE's are designed around local development how do you use them to collaborate a site, who names the variables and functions etc?

This never really seems clear to me!

polynamaude commented:
@Undefined-Variables I really understand nothing about your question....
**how do you use them to collaborate a site, who names the variables and functions etc?** ???

Give some example about what you mean here....
All programming language give name to the variables and function, that's how we reference them !

Undefined-Variables commented:
I'm sorry I was on phone so posted without reading back. 

So if most IDE software comes without the ability to share a remote server via ftp (upload and download on the fly) how would you collaborate on software? Do you have to rely on GitHub or if I had PHPStorm and you had PHPStorm is there a build in method to share work between two users? 

And second part, we say I will do the view and css you will do the controller and the model on a project. Where and how are the names of variables we use going to match? I might use $user_Name you might use $username. So I want to know how do we know what each other will use? 

This might seem like a stupid question but I have only ever worked alone so don't understand how it all comes together and fits?

polynamaude commented:
@Undefined-Variables That's a bit why I was saying that sometime it's important to pass by some part considered boring, that is algorithm development and structure.

Your question find it's answer in the use of documentation, standards, convention...
phpStorm, Eclipse, Visual Studio Code all manage the source code with the help of VCS / RCS (Version Control System / Revision Control System aka git, mercurial, svn). And yes even FTP ;-) I don't know when you looked but those are pretty much developed.

On a practical side, we mostly work with tags (example //TODO: fix this issue:111). There's whole documentation on this when you look on Gitlab.
You can try a free account (trial become free if you don't exceed 10 users) on Jira for project management. I think this will give you a good hint. Also, we use schematics based on UML as basis before coding, we make diagram.
We don't just write code, we make everything on flow chart beforehand. So all the variables are already decided. And even how to do it is coded. A simple rule for example would be that the visibility of a variable must be the lowest possible that permit it's function. So you can't make it _public_ if it can be done using _protected_, same you can't make it _protected_ if you can do it using _private_. 

Coding is not developing. It's one thing to make code that runs but to make code that last is something much harder. You don't want to end up having to write back a whole content management system because you forgot to make have different language for categories. You'll have already made it in a way that you just have to extend some classes and refer them...

If you have other question, I'll try to answer...

Undefined-Variables commented:
Oh wow that really makes sense and joint a lot of dots for me, I have wondered these things for years and assumed it had to be time staggered it was what made the most sense to me and asking questions like this make you look stupid in a professionals eyes, I never could fix one single job I wanted to do when I was growing up but if there was a way to travel back this is what I would do, I always liked lock and keys like padlocks and loved math and algebra and crosswords and puzzles but I was born in 1985 and none of this had evolved by time I was choosing a carrier path so I missed the time by a few years but later on in life I accidentally come across code and knew instantly I wanted to be involved as couldn't stop thinking about it. 

Thank you for taking the time to explain how this is done professionally it is really interesting :)

Undefined-Variables commented:
@polynamaude I sent a few hours looking at PHPStorm properly and can see the attraction for sure. This is like an all in one editor from the future, seriously it has everything you would ever want it’s quite incredible how smart it can be! 

Only dislike so far is CTRL & mouse wheel doesn’t zoom in/out... sometimes you just want to see the bigger picture or want to focus on one thing and I instantly missed it! 

Very nice. Thanks for tip :)

Undefined-Variables commented:
I saw when working in PHPStorm it creates the code for Android App but PHPStorm is unable to build apps, it suggests IDEA which is basically a clone of Android Studio. 

After install I retracted and renamed it and changed the url to the real one and spent 3hrs trying to figure out how to build but no matter what I tried I couldn’t work out how or where the module comes from, 85% of our users are on mobile so I need to either build app or pay for one to be built... Have you any experience using this or building apps in this way? If so do you mind explaining the basics to me? 

As you know knowing php css MySQL html does not help when it comes to apps written in Java it’s a different kettle of fish all together. A few pointers would be seriously appreciated mate.

polynamaude commented:
@Undefined-Variables Simply build a progressive web application. You'll get some offline possibility and a home screen icon.

https://developers.google.com/web/tools/workbox/guides/get-started

https://frontend.turing.io/lessons/module-4/pwa-intro-and-manifest.html

Undefined-Variables commented:
Wow thank you that is pretty cool :) Appreciated 👍

Undefined-Variables commented:
Yea I know what you mean by not just waking up and thinking today im going to do something I have no idea about. I'm under increasing pressure from users to get them an app. End users dont understand the difference they just want that website in an app! I was just reading you could write your app in JavaScript and use a program that converts in to Java in the background. This is the part about JavaScript that worries me "getting it wrong" cos nobody wants to get it wrong.

polynamaude commented:
@Undefined-Variables You should maybe take time to read what I write.... Or do it a bit better...
I never wrote something like you can write some application in Javascript and have it converted to Java...

What I did write is that you can write a PWA and this will give you the possibility to have you website with a icon on the user home screen or application list, be it a Android phone or any other system....
If you add a manifest to your site you'll give the user the possibility to have you web application look like a standard application (with limitation).

There's not really possibility to translate from JS to Java or the other way around...
I think a big problem you got is that you jump f*cking fast to conclusion and this doesn't only apply to  the computer side...

polynamaude commented:
@Undefined-Variables The main reason we use mathematics grade to evaluate admission in computer program is because mathematics are abstract and the inner working of computing is also pretty abstract thinking.

Mathematics is the mother of all science, well philosophy is the base of all this but mathematics can be seen as a branch of it. Mathematics is used to describe all the world around us and make rules pertaining to how it works.

So before all this maybe you better do so mental practice by doing back college level Algebra and this will make it easier to learn on from there when you start digging into programming language.

Now for the pressure coming from users, maybe you hadn't planned enough before diving into this. Like many person you had an idea and expected it to grow. That's nice but it has to be confronted to reality and put into context, planned, etc.

Now one of the thing that will be putting you most break when trying to have discussion with other programmer is that even if you have some knowledge, the simple fact that you use words in wrong context or to the wrond meaning will make look like a "script kiddie".

Take a break on Valium or whatever sleeping pills you take and your mind will see much clearer. Yes, it will be much a pain in the ass and it's awful to see the shit around us but you'll win on the long term.

Undefined-Variables commented:
No you must have lost something there, I said I was reading online that is one choice,. 

I understood what you said I have done the manifest and doing the service worker now.

polynamaude commented:
One example of what I said is this example of what's asynchronous code and you get it for browser side.

```
Window.onload = function() {
// When all requested document resources are loaded,
// do something with the resulting environment
}
element.onclick = function() {
// Do something when the user clicks on this element
}
```

If you want to learn JS then get Babel as processor. Not to transpile but to validate your code. You can also learn some Typescript as it's less error prone and total JS compatible.

polynamaude commented:
There's also callback (when you pass a function as a parameter and it's executed as part of the answer)

Undefined-Variables commented:
That’s interesting! I will definitely look into this! 

I have got pwa working perfectly now but had to put the manifest at the root as lighthouse wouldn’t recognise the start_url if it was done the way it was! 

Lighthouse now gives a green tick ✅ on everything! Now I have to work out how to get mobile phones to install it as android is not asking if. I want to add to home screen it just browse web like normal.

Undefined-Variables commented:
Understand now :)Thanks for advice!

Undefined-Variables commented:
@polynamaude Please mate would you stop bringing up things that I have said or done in the past? I have wiped the slate clean and wish to move on using this account, by bringing up things I have said in anger just makes me look stupid. You know what you are doing, we have talked enough that I know you are intelligent so you understand what I'm saying... Please I am asking you nicely, I thought we was past this silly behavior!

pH-7 commented:
Hey @Undefined-Variables 

I was just wondering if you got the chance to finish and update this one?


Thanks anyway!

Undefined-Variables commented:
Not yet mate, I am rewriting the IM module at the minute as this messenger although not quite working right it has the right idea, it has bugs though. Once I have this done I will start the PM system in a new folder/module called messages and apply the same concept as in IM module.

Once I am happy they can work along side each other with no bugs I will do PR :)

pH-7 commented:
Wow, that sounds really cool @Undefined-Variables! Can't wait to see it! :)
